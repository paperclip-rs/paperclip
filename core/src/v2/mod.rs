//! Core types and traits associated with the
//! [OpenAPI v2 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).

#[cfg(feature = "actix-base")]
mod actix;
mod extensions;
pub mod models;
#[cfg(feature = "codegen")]
mod resolver;
pub mod schema;

#[cfg(feature = "actix-base")]
pub use self::actix::{
    AcceptedJson, CreatedJson, NoContent, OperationModifier, ResponderWrapper, ResponseWrapper,
};

pub use self::models::{DefaultSchema, ResolvableApi};
pub use self::schema::Schema;
pub use paperclip_macros::*;

#[cfg(feature = "codegen")]
use self::resolver::Resolver;
#[cfg(feature = "codegen")]
use crate::error::ValidationError;

#[cfg(feature = "codegen")]
impl<S: Schema + Default> ResolvableApi<S> {
    /// Consumes this API schema, resolves the references and returns
    /// the resolved schema.
    ///
    /// This walks recursively, collects the referenced schema objects,
    /// substitutes the referenced IDs with the pointer to schema objects
    /// and returns the resolved object or an error if it encountered one.
    pub fn resolve(self) -> Result<ResolvableApi<S>, ValidationError> {
        let mut resolver = Resolver::from((
            self.definitions,
            self.paths,
            self.parameters,
            self.responses,
        ));
        resolver.resolve()?;
        Ok(ResolvableApi {
            swagger: self.swagger,
            info: self.info,
            definitions: resolver.defs,
            paths: resolver.paths,
            base_path: self.base_path,
            host: self.host,
            schemes: self.schemes,
            consumes: self.consumes,
            produces: self.produces,
            coders: self.coders,
            support_crates: self.support_crates,
            parameters: resolver.params,
            responses: resolver.resp,
            spec_format: self.spec_format,
            external_docs: self.external_docs,
            security: self.security,
            security_definitions: self.security_definitions,
            tags: self.tags,
        })
    }
}
