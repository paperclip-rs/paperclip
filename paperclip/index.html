<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Paperclip is a OpenAPI code generator for efficient type-safe compile-time checked HTTP APIs in Rust."><title>paperclip - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="paperclip" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0-nightly (f3db63916 2025-06-17)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../paperclip/index.html">paperclip</a><span class="version">0.9.6</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#modules">Crate Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>paperclip</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/paperclip/lib.rs.html#1-34">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Paperclip is a OpenAPI code generator for efficient type-safe
compile-time checked HTTP APIs in Rust.</p>
<p>See the <a href="https://paperclip-rs.github.io/paperclip">website</a> for detailed
documentation and examples.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="actix/index.html" title="mod paperclip::actix">actix</a></dt><dd>Plugin types, traits and macros for actix-web framework.</dd><dt><a class="mod" href="util/index.html" title="mod paperclip::util">util</a></dt><dt><a class="mod" href="v2/index.html" title="mod paperclip::v2">v2</a></dt><dd>Utilities related to the <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md">OpenAPI v2 specification</a>.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.PaperClipError.html" title="enum paperclip::PaperClipError">Paper<wbr>Clip<wbr>Error</a></dt><dd>Global error which encapsulates all related errors.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="type" href="type.PaperClipResult.html" title="type paperclip::PaperClipResult">Paper<wbr>Clip<wbr>Result</a></dt><dd>Generic result used throughout this library.</dd></dl><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="attr" href="attr.api_v2_schema.html" title="attr paperclip::api_v2_schema">api_<wbr>v2_<wbr>schema</a></dt><dd>Converts your struct to support deserializing from an OpenAPI v2
<a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject">Schema</a>
object (<a href="https://paperclip-rs.github.io/paperclip/paperclip/v2/">example</a>). This adds the necessary fields (in addition to your own fields) and implements the
<code>Schema</code> trait for parsing and codegen.</dd></dl></section></div></main></body></html>