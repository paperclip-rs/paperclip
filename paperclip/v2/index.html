<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Utilities related to the OpenAPI v2 specification."><title>paperclip::v2 - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="paperclip" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0-nightly (f3db63916 2025-06-17)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../paperclip/index.html">paperclip</a><span class="version">0.9.5</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module v2</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#detailed-example" title="Detailed example">Detailed example</a></li></ul><h3><a href="#reexports">Module Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate paperclip</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">paperclip</a></div><h1>Module <span>v2</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/paperclip/v2/mod.rs.html#1-164">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Utilities related to the <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md">OpenAPI v2 specification</a>.</p>
<h2 id="detailed-example"><a class="doc-anchor" href="#detailed-example">§</a>Detailed example</h2>
<p>To parse your v2 spec, you begin with transforming the schema into a
Rust struct. If your schema doesn’t have custom properties, then you
can use the <code>DefaultSchema</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>paperclip::v2::{<span class="self">self</span>, ResolvableApi, DefaultSchema, models::Version};

<span class="kw">use </span>std::fs::File;

<span class="kw">let </span><span class="kw-2">mut </span>fd = File::open(<span class="string">"my_spec.yaml"</span>).unwrap(); <span class="comment">// yaml or json
</span><span class="kw">let </span>api: ResolvableApi&lt;DefaultSchema&gt; = v2::from_reader(<span class="kw-2">&amp;mut </span>fd).unwrap();
<span class="macro">assert_eq!</span>(api.swagger, Version::V2);</code></pre></div>
<p>On the other hand, if your schema does have custom properties which you’d
like to parse, then use the <code>#[api_v2_schema]</code> proc macro.</p>
<p>For example, let’s take the <a href="https://github.com/kubernetes/kubernetes/tree/afd928b8bc81cea385eba4c94558373df7aeae75/api/openapi-spec">Kubernetes API spec</a>
which uses some custom thingmabobs. Let’s say we’re only interested in the
<code>x-kubernetes-patch-strategy</code> field for now.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[macro_use] </span><span class="kw">extern crate </span>serde_derive; <span class="comment">// NOTE: We're using serde for decoding stuff.

</span><span class="kw">use </span>paperclip::v2::{<span class="self">self</span>, ResolvableApi};

<span class="kw">use </span>std::fs::File;

<span class="attr">#[derive(Debug, Deserialize)]
#[serde(rename_all = <span class="string">"camelCase"</span>)]
</span><span class="kw">enum </span>PatchStrategy {
    Merge,
    RetainKeys,
    <span class="attr">#[serde(rename = <span class="string">"merge,retainKeys"</span>)]
    </span>MergeAndRetain,
    <span class="attr">#[serde(other)]
    </span>Other,
}

<span class="attr">#[paperclip::api_v2_schema]
#[derive(Debug, Deserialize)]
</span><span class="kw">struct </span>K8sSchema {
    <span class="attr">#[serde(rename = <span class="string">"x-kubernetes-patch-strategy"</span>)]
    </span>patch_strategy: <span class="prelude-ty">Option</span>&lt;PatchStrategy&gt;,
}

<span class="comment">// K8sSchema now implements `Schema` trait.
</span><span class="kw">let </span><span class="kw-2">mut </span>fd = File::open(<span class="string">"k8s_spec.yaml"</span>).unwrap();
<span class="kw">let </span>api: ResolvableApi&lt;K8sSchema&gt; = v2::from_reader(<span class="kw-2">&amp;mut </span>fd).unwrap();</code></pre></div>
<p>Now, if <code>codegen</code> feature is enabled (it is by default), we can use the
emitter to emit the API into some path. But first, we need to resolve the
raw schema. During resolution, we:</p>
<ul>
<li>walk through the nodes, find <code>$ref</code> fields and assign references to
the actual definitions.</li>
<li>identify anonymous definitions in body parameters and response schemas
and add them to the known map of definitions.</li>
</ul>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[cfg(feature = <span class="string">"codegen"</span>)] </span>{

<span class="kw">let </span>resolved = api.resolve().unwrap();
}</code></pre></div>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[cfg(feature = <span class="string">"codegen"</span>)] </span>{
<span class="kw">use </span>paperclip::v2::{DefaultEmitter, EmitterState, Emitter};

<span class="kw">let </span><span class="kw-2">mut </span>state = EmitterState::default();
state.working_dir = <span class="string">"/path/to/my/crate"</span>.into();
<span class="kw">let </span>emitter = DefaultEmitter::from(state);
emitter.generate(<span class="kw-2">&amp;</span>api).unwrap(); <span class="comment">// generate code!
</span>}</code></pre></div>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><dl class="item-table reexports"><dt id="reexport.DefaultEmitter"><code>pub use self::codegen::<a class="struct" href="codegen/struct.DefaultEmitter.html" title="struct paperclip::v2::codegen::DefaultEmitter">DefaultEmitter</a>;</code></dt><dt id="reexport.Emitter"><code>pub use self::codegen::<a class="trait" href="codegen/trait.Emitter.html" title="trait paperclip::v2::codegen::Emitter">Emitter</a>;</code></dt><dt id="reexport.EmitterState"><code>pub use self::codegen::<a class="struct" href="codegen/struct.EmitterState.html" title="struct paperclip::v2::codegen::EmitterState">EmitterState</a>;</code></dt><dt id="reexport.serde_json"><code>pub use paperclip_core::v2::<a class="mod" href="https://docs.rs/serde_json/1.0.140/serde_json/index.html" title="mod serde_json">serde_json</a>;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="codegen/index.html" title="mod paperclip::v2::codegen">codegen</a></dt><dd>Code generation for OpenAPI v2.</dd><dt><a class="mod" href="im/index.html" title="mod paperclip::v2::im">im</a></dt><dd>Interior mutability stuff.</dd><dt><a class="mod" href="models/index.html" title="mod paperclip::v2::models">models</a></dt><dd>Models used by OpenAPI v2.</dd><dt><a class="mod" href="schema/index.html" title="mod paperclip::v2::schema">schema</a></dt><dd>Traits used for code and spec generation.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.DefaultSchema.html" title="struct paperclip::v2::DefaultSchema">Default<wbr>Schema</a></dt><dd>Default schema if your schema doesn’t have any custom fields.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Schema.html" title="trait paperclip::v2::Schema">Schema</a></dt><dd>Interface for the <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject"><code>Schema</code></a> object.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.from_reader.html" title="fn paperclip::v2::from_reader">from_<wbr>reader</a></dt><dd>Deserialize the schema from the given reader. Currently, this only supports
JSON and YAML formats.</dd><dt><a class="fn" href="fn.from_reader_v3.html" title="fn paperclip::v2::from_reader_v3">from_<wbr>reader_<wbr>v3</a></dt><dd>Deserialize the schema from the given reader. Currently, this only supports
JSON and YAML formats.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.ResolvableApi.html" title="type paperclip::v2::ResolvableApi">Resolvable<wbr>Api</a></dt><dd>OpenAPI v2 spec which can be traversed and resolved for codegen.</dd></dl></section></div></main></body></html>