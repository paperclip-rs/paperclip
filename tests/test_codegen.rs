use paperclip::v2::{
    self,
    codegen::{CrateMeta, DefaultEmitter, EmitMode, Emitter, EmitterState},
    models::{DefaultSchema, ResolvableApi},
};

use once_cell::sync::Lazy;
use std::{fs::File, path::PathBuf};

static ROOT: Lazy<PathBuf> = Lazy::new(|| PathBuf::from(env!("CARGO_MANIFEST_DIR")));
static PET_SCHEMA: Lazy<ResolvableApi<DefaultSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/pet-v2.yaml")).expect("file?");
    let raw: ResolvableApi<DefaultSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static K8S_SCHEMA: Lazy<ResolvableApi<K8sSchema>> = Lazy::new(|| {
    let fd = File::open(ROOT.join("tests/k8s-v1.16.0-alpha.0-openapi-v2.json")).expect("file?");
    let raw: ResolvableApi<K8sSchema> = v2::from_reader(fd).expect("deserializing spec");
    raw.resolve().expect("resolution")
});
static CODEGEN_PET_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_LIB_NO_ROOT: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = ROOT.clone();
    state.working_dir.push("tests/test_pet/no_root");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::Crate;
    meta.no_root = true;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});
static CODEGEN_PET_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_pet/cli");
    let mut meta = CrateMeta::default();
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&PET_SCHEMA).expect("codegen");
});

static CODEGEN_K8S_LIB: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests");
    state.working_dir.push("test_k8s");
    state.mod_prefix = "crate::codegen::";

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});
static CODEGEN_K8S_CLI: Lazy<()> = Lazy::new(|| {
    let mut state = EmitterState::default();
    state.working_dir = (&*ROOT).into();
    state.working_dir.push("tests/test_k8s/cli");
    let mut meta = CrateMeta::default();
    assert_eq!(meta.mode, EmitMode::Module);
    meta.name = Some("test-k8s-cli".into());
    meta.version = Some("0.0.0".into());
    meta.authors = Some(vec!["Me <me@example.com>".into()]);
    meta.mode = EmitMode::App;
    state.set_meta(meta);

    let emitter = DefaultEmitter::from(state);
    emitter.generate(&K8S_SCHEMA).expect("codegen");
});

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
enum PatchStrategy {
    Merge,
    RetainKeys,
    #[serde(rename = "merge,retainKeys")]
    MergeAndRetain,
}

#[paperclip::api_v2_schema]
#[allow(dead_code)]
#[derive(Debug, serde::Deserialize)]
struct K8sSchema {
    #[serde(rename = "x-kubernetes-patch-strategy")]
    patch_strategy: Option<PatchStrategy>,
}

static CODEGEN: Lazy<()> = Lazy::new(|| {
    env_logger::builder()
        .filter(Some("paperclip"), log::LevelFilter::Info)
        .init();
    Lazy::force(&CODEGEN_PET_LIB);
    Lazy::force(&CODEGEN_PET_LIB_NO_ROOT);
    Lazy::force(&CODEGEN_PET_CLI);
    Lazy::force(&CODEGEN_K8S_LIB);
    Lazy::force(&CODEGEN_K8S_CLI);
});

fn assert_file(path: &str) {
    Lazy::force(&CODEGEN);

    let data = std::fs::read_to_string(ROOT.join(path))
        .unwrap_or_else(|err| panic!("missing file {}: {}", path, err));
    insta::assert_snapshot!(path, data);
}

#[cfg(test)]
mod tests_pet {
    use super::*;
    #[test]
    fn test_lib_root() {
        assert_file("tests/test_pet/lib.rs");
    }

    #[test]
    fn test_no_root_manifest() {
        assert_file("tests/test_pet/no_root/Cargo.toml");
    }

    #[test]
    fn test_manifest() {
        assert_file("tests/test_pet/Cargo.toml");
    }

    #[test]
    fn test_header_parameters() {
        assert_file("tests/test_pet/pet.rs");
    }

    #[test]
    fn test_cli_params() {
        assert_file("tests/test_pet/cli/app.yaml");
    }

    #[test]
    fn test_recusrive_object() {
        assert_file("tests/test_pet/recursive_container.rs");
        assert_file("tests/test_pet/recursive_object.rs");
    }

    #[test]
    fn test_anonymous_object_definition_in_schema() {
        // An object "can" define objects in its schema without referencing
        // them from known definitions. In that case, we autogenerate stuff.
        assert_file("tests/test_pet/order.rs");
    }
    #[test]
    fn test_anonymous_object_definition_in_body() {
        assert_file("tests/test_pet/post_shipments_body.rs");
    }

    #[test]
    fn test_anonumous_object_definition_in_response() {
        assert_file("tests/test_pet/get_shipments_id_response.rs");
    }

    #[test]
    fn test_simple_array_parameter_in_path() {
        assert_file("tests/test_pet/status.rs");
    }

    #[test]
    fn test_misc() {
        assert_file("tests/test_pet/miscellaneous.rs");
    }

    #[test]
    fn test_multipart_with_file() {
        assert_file("tests/test_pet/cli/status.rs");
    }

    #[test]
    fn test_simple_any_enum() {
        assert_file("tests/test_pet/test_enum.rs");
    }
}

#[cfg(test)]
mod tests_k8s {
    use super::*;
    use paperclip::v2::models::{HttpMethod, Version};

    #[test]
    fn test_definition_ref_cycles() {
        assert_eq!(K8S_SCHEMA.swagger, Version::V2);
        assert_eq!(K8S_SCHEMA.definitions.len(), 663);

        let json_props_def = &K8S_SCHEMA.definitions
            ["io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps"];
        let desc = json_props_def.read().description.clone();
        let all_of = json_props_def.read().properties["allOf"].clone();
        let items = all_of.read().items.as_ref().unwrap().clone();
        assert_eq!(items.read().description, desc); // both point to same `JSONSchemaProps`
    }

    #[test]
    fn test_resolved_schema() {
        let resp = &K8S_SCHEMA.paths["/api/"].methods[&HttpMethod::Get].responses["200"].read();
        let schema = resp.schema.as_ref().expect("bleh?").read();
        assert!(schema.reference.is_none()); // this was a reference
        assert_eq!(
            &K8S_SCHEMA.definitions["io.k8s.apimachinery.pkg.apis.meta.v1.APIVersions"]
                .read()
                .description
                .as_ref()
                .unwrap()
                .as_str(),
            schema.description.as_ref().unwrap()
        );
    }

    #[test]
    fn test_child_module_declarations() {
        assert_file("tests/test_k8s/io/k8s/api/mod.rs");
    }

    #[test]
    fn test_transparency_with_params() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/custom_resource_definition.rs");
    }
    #[test]
    fn test_struct_for_complex_object() {
        assert_file("tests/test_k8s/io/k8s/apiextensions_apiserver/pkg/apis/apiextensions/v1beta1/json_schema_props.rs");
    }

    #[test]
    fn test_root_mod() {
        assert_file("tests/test_k8s/mod.rs");
    }

    #[test]
    fn test_generics_mod() {
        assert_file("tests/test_k8s/generics.rs");
    }

    #[test]
    fn test_same_object_creates_multiple_builders() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/config_map.rs")
    }

    #[test]
    fn test_same_object_with_multiple_builders_has_basic_builder() {
        assert_file("tests/test_k8s/io/k8s/api/core/v1/pod.rs")
    }

    #[test]
    fn test_simple_object_builder_with_required_fields() {
        assert_file("tests/test_k8s/io/k8s/api/rbac/v1/policy_rule.rs")
    }

    #[test]
    fn test_builder_with_field_parameter_collision_and_method_collision() {
        // grace_period_seconds, orphan_dependents and propagation_policy
        // exist in both the object and as a query parameter. If one is set,
        // we should also set the other.
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/delete_options.rs");
    }

    #[test]
    fn test_unit_builder_with_no_modifier() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/api_group_list.rs");
    }

    #[test]
    fn test_builder_field_with_iterators() {
        assert_file(
            "tests/test_k8s/io/k8s/api/certificates/v1beta1/certificate_signing_request_spec.rs",
        );
    }

    #[test]
    fn test_any_in_operation_bound_to_unrelated_struct() {
        assert_file("tests/test_k8s/io/k8s/apimachinery/pkg/apis/meta/v1/patch.rs");
    }

    #[test]
    fn test_cli_manifest() {
        assert_file("tests/test_k8s/cli/Cargo.toml");
    }

    #[test]
    fn test_cli_main() {
        assert_file("tests/test_k8s/cli/main.rs");
    }

    #[test]
    fn test_clap_yaml() {
        assert_file("tests/test_k8s/cli/app.yaml");
    }
}
