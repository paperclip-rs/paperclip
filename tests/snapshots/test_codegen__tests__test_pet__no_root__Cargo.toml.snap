---
source: tests/test_codegen.rs
expression: data
---
[package]
name = "no_root"
version = "0.1.0"
authors = ["Me <me@example.com>"]
edition = "2018"

[lib]
path = "lib.rs"

[dependencies]
async-trait = "0.1"
bytes = "0.5"
thiserror = "1.0"
futures = "0.3"
http = "0.2"
lazy_static = "1.4"
log = "0.4"
mime = { git = "https://github.com/hyperium/mime" }
mime_guess = "2.0"
parking_lot = "0.11"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
serde_yaml = "0.8"
tokio-util = { version = "0.4", features = ["codec"] }
url = "2.1"

tokio = { version = "0.3", features = ["fs", "io-util"] }
reqwest = { version = "0.10", features = ["stream", "json"] }



